src/safepoint.c

```diff
@@ -276,6 +276,21 @@ void jl_safepoint_wait_thread_resume(void)
     jl_atomic_store_release(&ct->ptls->gc_state, state);
     uv_mutex_unlock(&ct->ptls->sleep_lock);
 }
+// This takes the sleep lock and puts the thread in GC_SAFE
+int8_t jl_safepoint_take_sleep_lock(jl_ptls_t ptls)
+{
+    int8_t gc_state = jl_gc_safe_enter(ptls);
+    uv_mutex_lock(&ptls->sleep_lock);
+    if (jl_atomic_load_relaxed(&ptls->suspend_count)) {
+        // defer this broadcast until we determine whether uv_cond_wait is really going to be needed
+        uv_mutex_unlock(&ptls->sleep_lock);
+        uv_mutex_lock(&safepoint_lock);
+        uv_cond_broadcast(&safepoint_cond_begin);
+        uv_mutex_unlock(&safepoint_lock);
```

Comment:
Could you explain more what the actual deadlock is? I don't fully understand this code, but this point here where no lock are held is suspicious to me, because I don't understand why it's any different than the state before. What prevents whatever bad condition what happen before from happening again in between this unlock and the subsequent lock?

Reply:
@vtjnash Does that look right to you?